File: python/04_queue/array_queue.py
# coding:utf-8

""""
用数组实现队列
"""


class ArrayQueue:

    def __init__(self, capacity: int):
        self._items = []
        self._capacity = capacity
        self._head = 0
        self._tail = 0

    def enqueue(self, item: str) -> bool:
        if self._tail == self._capacity:
            if self._head == 0:
                return False
            else:  # 队列中前面有空的位置,进行数据搬移
                for i in range(0, self._tail - self._head):
                    self._items[i] = self._items[i + self._head]
                # 搬移完成后，修改头尾指针指向
                self._tail -= self._head
                self._head = 0
        self._items.insert(self._tail, item)
        self._tail += 1
        return True

    def dequeue(self) -> str:
        if self._tail != self._head:
            item = self._items[self._head]
            self._head += 1
            return item
        else:
            return None

    def __repr__(self) -> str:
        return " ".join(item for item in self._items[self._head:self._tail])


if __name__ == '__main__':
    array_queue = ArrayQueue(3)
    print(array_queue.enqueue('1'))
    print(array_queue.enqueue('2'))
    print(array_queue.enqueue('3'))
    print(array_queue.enqueue('4'))
    print(array_queue)
    print(array_queue.dequeue())
    print(array_queue)
    print(array_queue.dequeue())
    print(array_queue.dequeue())
    print(array_queue)

File: go/04_queue/array_queue.go
package _4_queue

import (
	"fmt"
)

// 基于数组的队列
type ArrayQueue struct {
	data     []interface{}
	capacity int
	head     int
	tail     int
}

func (q *ArrayQueue) EnQueue(v interface{}) bool {
	if q.tail == q.capacity {
		if q.head == 0 {
			return false
		} else { // 队列中前面有空的位置,进行数据搬移

			for i := 0; i < q.tail-q.head; i++ {
				q.data[i] = q.data[i+q.head]
			}
			// 搬移完成后，修改头尾指针指向
			q.tail -= q.head
			q.head = 0
		}

	}

	q.data[q.tail] = v
	q.tail++
	return true
}

func (q *ArrayQueue) DeQueue() interface{} {
	if q.head == q.capacity {
		return nil
	}
	v := q.data[q.head]
	q.head++
	return v
}

func (q *ArrayQueue) Print() string {


	result := ""
	for i := q.head; i < q.tail-1; i++ {
		result += fmt.Sprintf("%v->", q.data[i])
	}
	result += fmt.Sprintf("%v", q.data[q.tail-1])
	return result


}

func NewArrayQueue(n int) *ArrayQueue {
	return &ArrayQueue{make([]interface{}, n), n, 0, 0}
}
